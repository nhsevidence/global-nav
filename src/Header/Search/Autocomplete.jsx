import React, { Component } from "react";
import Autocomplete from "accessible-autocomplete/react";

import "!style-loader!css-loader!accessible-autocomplete/dist/accessible-autocomplete.min.css";

function suggest(query, syncResults) {
	fetch(`https://www.nice.org.uk/autocomplete?q=${query}&ajax=ajax`)
		.then(response => response.json())
		.then(data => {
			syncResults(data);
		});
}

const templates = {
	inputValue: suggestion => suggestion && suggestion.Title,
	suggestion: suggestion =>
		suggestion && `<a href="${suggestion.Link}">${suggestion.Title}</a>`
};

const onConfirm = suggestion => {
	if (suggestion) window.location.href = suggestion.Link;
};

export default class Search extends Component {
	render() {
		return (
			<Autocomplete
				id="autocomplete"
				source={suggest}
				templates={templates}
				onConfirm={onConfirm}
			/>
		);
	}
}
