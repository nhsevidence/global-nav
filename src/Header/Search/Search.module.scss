@import '~@nice-digital/design-system/src/stylesheets/settings/settings';
@import '~@nice-digital/design-system/src/stylesheets/helpers/helpers-utils';

$size: rem(40);

.search {
  padding-right: $size;
  position: relative;
}

.label {
  /* https://snook.ca/archives/html_and_css/hiding-content-for-accessibility */
  // TO DO: Replace with design system invisible mixin when available
  clip: rect(1px 1px 1px 1px); /* IE6, IE7 */
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute !important;
  width: 1px;
}

.button {
  background: $colour-nice-dark-grey;
  border: 0;
  color: $colour-text-inverse;
  cursor: pointer;
  height: $size;
  min-width: $size;
  padding: rem(0 $spacing-small);
  position: absolute;
  right: 0;
  text-align: center;
  top: 0;

  &:focus {
    // TODO:  Replace with design system mixin
    outline: 3px solid $colour-focus;
  }

  .icon {
    font-size: 150%;
    vertical-align: middle;
  }
}
