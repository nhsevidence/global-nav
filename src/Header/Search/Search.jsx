import React, { Component } from "react";
import PropTypes from "prop-types";
import SearchIcon from "@nice-digital/icons/lib/Search";

import Autocomplete from "./Autocomplete";

import styles from "./Search.module.scss";

export default class Search extends Component {
	constructor(props) {
		super(props);

		this.searchSubmitHandler = this.searchSubmitHandler.bind(this);
	}

	searchSubmitHandler(e) {
		const { onSearching } = this.props;
		if (onSearching) {
			const onSearchingCallback =
				typeof onSearching === "function" ? onSearching : window[onSearching];

			if (typeof onSearchingCallback === "function") {
				e.preventDefault();
				const query = document.getElementById("autocomplete").value;
				onSearchingCallback({ query: query });
			}
		}
	}

	render() {
		return (
			<form
				role="search"
				action={this.props.url}
				className={styles.search}
				onSubmit={this.searchSubmitHandler}
			>
				{/* eslint jsx-a11y/label-has-for: 0 */}
				<label className={styles.label} htmlFor="autocomplete">
					{this.props.placeholder}
				</label>
				<Autocomplete
					source={this.props.autocomplete}
					placeholder={this.props.placeholder}
					query={this.props.query}
				/>
				<button
					className={styles.button}
					type="submit"
					aria-label="Perform search"
				>
					{!window || typeof window.SVGRect !== "undefined" ? (
						<SearchIcon className={styles.icon} />
					) : (
						// We can remove this fallback when we drop support for IE8
						<span className={styles.icon} aria-hidden="true">
							search
						</span>
					)}
				</button>
			</form>
		);
	}
}

Search.propTypes = {
	url: PropTypes.string,
	autocomplete: PropTypes.oneOfType([
		PropTypes.bool,
		PropTypes.string,
		PropTypes.arrayOf(
			PropTypes.shape({ Title: PropTypes.string, Link: PropTypes.string })
		)
	]),
	placeholder: PropTypes.string,
	query: PropTypes.string,
	onSearching: PropTypes.string
};

Search.defaultProps = {
	url: "/search",
	placeholder: "Search NICEâ€¦",
	query: ""
};
