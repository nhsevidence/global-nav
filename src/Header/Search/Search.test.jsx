import React from "react";
import Search from "./Search";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

describe("Search", () => {
	const defaultProps = {};

	describe("Rendering", () => {
		it("Renders without crashing", () => {
			const wrapper = shallow(<Search {...defaultProps} />);
			expect(wrapper).toHaveLength(1);
		});

		it("Matches snapshot", () => {
			const wrapper = shallow(<Search {...defaultProps} />);
			expect(toJson(wrapper)).toMatchSnapshot();
		});

		it("Renders SVG search icon when in an SVG-compatible browser", () => {
			window.SVGRect = {};
			const wrapper = shallow(<Search {...defaultProps} />);
			expect(wrapper.find("SvgSearch").length).toEqual(1);
			window.SVGRect = undefined;
		});

		it("Renders fallback unicode search icon when in an non SVG-compatible browser", () => {
			const wrapper = shallow(<Search {...defaultProps} />);
			expect(wrapper.find("button").text()).toEqual("âŒ•");
		});
	});

	describe("Autocomplete", () => {
		it("Passes props to autocomplete component", () => {
			const wrapper = shallow(
				<Search
					{...defaultProps}
					autocomplete="variableName"
					placeholder="Test placeholder"
				/>
			);
			expect(wrapper.find("Autocomplete").props()).toEqual({
				source: "variableName",
				placeholder: "Test placeholder"
			});
		});
	});
});
