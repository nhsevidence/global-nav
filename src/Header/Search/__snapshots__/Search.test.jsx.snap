// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Rendering Matches snapshot 1`] = `
<form
  action="/search"
  className="search"
  onSubmit={[Function]}
  role="search"
>
  <label
    className="label"
    htmlFor="autocomplete"
  >
    Search NICE…
  </label>
  <Autocomplete
    placeholder="Search NICE…"
    query=""
    source={false}
  />
  <button
    aria-label="Perform search"
    className="button"
    type="submit"
  >
    <span
      aria-hidden="true"
      className="icon"
    >
      search
    </span>
  </button>
</form>
`;
